[ecovacs] Robot not updated after some time

<!-- Provide a general summary of the issue in the *Title* above -->
<!-- If the issue is related to a binding, please include its short name in -->
<!-- square brackets in the title - Example: "[astro] My issue..." -->

<!-- Important: Please contact the openHAB community forum for questions or -->
<!-- for configuration and usage guidance: https://community.openhab.org -->

<!-- Feel free to delete any comment lines in the template (starting with "<!--") -->

## Expected Behavior
<!-- If you're describing a bug, tell us what should happen -->
<!-- If you're suggesting a change/improvement, tell us how it should work -->

As long as the things for the API account and the robot (DEEBOT OZMO 900 Series) are online, sending a command should work.

## Current Behavior
<!-- If describing a bug, tell us what happens instead of the expected behavior -->
<!-- Include related log information (preferably debug level) and related configs -->
<!-- Use a file attachment for log and config information longer than a few lines -->
<!-- Enclose multi-line log/code snippets with ``` on new lines for proper formatting -->
<!-- If suggesting a change/improvement, explain the difference from current behavior -->
<!-- For improvements, discuss at community.openhab.org first and include link to topic -->

After some days the robot cannot be controlled via openHAB anymore and passive data isn't updated anymore (e.g. current cleaning time). Both the account and the bot are still online in openHAB and in the official Ecovacs app.

In the log I can see the following lines:
````
10:31:08.835 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel total-stats#total-cleaned-area although the handler was already disposed.
10:31:08.837 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel total-stats#total-cleaning-time although the handler was already disposed.
10:31:08.839 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel total-stats#total-clean-runs although the handler was already disposed.
10:31:08.941 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel settings#continuous-cleaning although the handler was already disposed.
10:31:09.045 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel last-clean#last-clean-start although the handler was already disposed.
10:31:09.046 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel last-clean#last-clean-duration although the handler was already disposed.
10:31:09.050 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel last-clean#last-clean-area although the handler was already disposed.
10:31:10.492 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel settings#suction-power although the handler was already disposed.
10:31:10.578 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel settings#water-amount although the handler was already disposed.
10:31:10.683 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel consumables#side-brush-lifetime although the handler was already disposed.
10:31:10.780 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel consumables#dust-filter-lifetime although the handler was already disposed.
10:31:10.887 [WARN ] [b.core.thing.binding.BaseThingHandler] - Handler EcovacsVacuumHandler of thing ecovacs:vacuum:2b4b2f4b73:E0001206219602620054 tried updating channel consumables#main-brush-lifetime although the handler was already disposed.
````

## Possible Solution
<!-- Not obligatory, but suggest a fix/reason for the bug, -->
<!-- or ideas how to implement the addition or change -->

## Steps to Reproduce (for Bugs)
<!-- Provide a link to a live example, or an unambiguous set of steps to -->
<!-- reproduce this bug. Include code to reproduce, if relevant -->
1. Wait for a few days

## Context
<!-- How has this issue affected you? What are you trying to accomplish? -->
<!-- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!-- Include as many relevant details about the environment you experienced the bug in -->
* Version used: (e.g., openHAB and add-on versions) Docker container `openhab/openhab:4.0.4`
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Raspbian Bullseye

Issue appeared again and I got this exception:

````
[DEBUG] [internal.handler.EcovacsVacuumHandler] - E0001206219602620054: Failed communicating to device, reconnecting
org.openhab.binding.ecovacs.internal.api.EcovacsApiException: java.util.concurrent.ExecutionException: java.util.concurrent.RejectedExecutionException: HttpClient@270fb9bd{STOPPED} is stopped
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsApiImpl.executeRequest(EcovacsApiImpl.java:391) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsApiImpl.sendIotCommand(EcovacsApiImpl.java:289) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsIotMqDevice.sendCommand(EcovacsIotMqDevice.java:97) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsIotMqDevice.connect(EcovacsIotMqDevice.java:123) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.handler.EcovacsVacuumHandler.lambda$14(EcovacsVacuumHandler.java:561) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.handler.EcovacsVacuumHandler.doWithDevice(EcovacsVacuumHandler.java:798) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.handler.EcovacsVacuumHandler.connectToDevice(EcovacsVacuumHandler.java:560) ~[bundleFile:?]
        at org.openhab.binding.ecovacs.internal.api.util.SchedulerTask.run(SchedulerTask.java:95) [bundleFile:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.RejectedExecutionException: HttpClient@270fb9bd{STOPPED} is stopped
        at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:732) ~[bundleFile:9.4.52.v20230823]
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsApiImpl.executeRequest(EcovacsApiImpl.java:385) ~[bundleFile:?]
        ... 13 more
Caused by: java.util.concurrent.RejectedExecutionException: HttpClient@270fb9bd{STOPPED} is stopped
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:355) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:323) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:627) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpRequest.sendAsync(HttpRequest.java:780) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:767) ~[bundleFile:9.4.52.v20230823]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:728) ~[bundleFile:9.4.52.v20230823]
        at org.openhab.binding.ecovacs.internal.api.impl.EcovacsApiImpl.executeRequest(EcovacsApiImpl.java:385) ~[bundleFile:?]
        ... 13 more
````

--- a/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/AbstractPortalIotCommandResponse.java
+++ b/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/AbstractPortalIotCommandResponse.java
@@ -24,14 +24,12 @@ public class AbstractPortalIotCommandResponse {
     @SerializedName("errno")
     private final int errorCode;
     @SerializedName("error")
-    private final String errorMessage;
+    private final Object errorObject; // might be a string or a JSON object
 
-    // unused field: 'id' (string)
-
-    public AbstractPortalIotCommandResponse(String result, int errorCode, String errorMessage) {
+    public AbstractPortalIotCommandResponse(String result, int errorCode, Object errorObject) {
         this.result = result;
         this.errorCode = errorCode;
-        this.errorMessage = errorMessage;
+        this.errorObject = errorObject;
     }
 
     public boolean wasSuccessful() {
@@ -39,13 +37,22 @@ public class AbstractPortalIotCommandResponse {
     }
 
     public boolean failedDueToAuthProblem() {
-        return "fail".equals(result) && errorMessage != null && errorMessage.toLowerCase().contains("auth error");
+        if (!"fail".equals(result)) {
+            return false;
+        }
+        if (errorCode == 3) {
+            // Error 3 is 'OAuth error'
+            return true;
+        }
+        String errorMessage = errorObject != null ? errorObject.toString().toLowerCase() : "";
+        return errorMessage.contains("auth error") || errorMessage.contains("token error");
     }
 
     public String getErrorMessage() {
         if (wasSuccessful()) {
             return null;
         }
+        String errorMessage = errorObject != null ? errorObject.toString() : null;
         return "result=" + result + ", errno=" + errorCode + ", error=" + errorMessage;
     }
 }

--- a/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/PortalIotCommandJsonResponse.java
+++ b/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/PortalIotCommandJsonResponse.java
@@ -27,8 +27,8 @@ public class PortalIotCommandJsonResponse extends AbstractPortalIotCommandRespon
     @SerializedName("resp")
     public final JsonElement response;
 
-    public PortalIotCommandJsonResponse(String result, JsonElement response, int errorCode, String errorMessage) {
-        super(result, errorCode, errorMessage);
+    public PortalIotCommandJsonResponse(String result, JsonElement response, int errorCode, Object errorObject) {
+        super(result, errorCode, errorObject);
         this.response = response;
     }
 

--- a/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/PortalIotCommandXmlResponse.java
+++ b/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/dto/response/portal/PortalIotCommandXmlResponse.java
@@ -21,8 +21,8 @@ public class PortalIotCommandXmlResponse extends AbstractPortalIotCommandRespons
     @SerializedName("resp")
     private final String responseXml;
 
-    public PortalIotCommandXmlResponse(String result, String responseXml, int errorCode, String errorMessage) {
-        super(result, errorCode, errorMessage);
+    public PortalIotCommandXmlResponse(String result, String responseXml, int errorCode, Object errorObject) {
+        super(result, errorCode, errorObject);
         this.responseXml = responseXml;
     }
 

